runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java

```diff
@@ -94,20 +95,29 @@
                 .getUnderlyingObject()
                 .access(o -> {
                             WriteSetSMRStream stream = o.getOptimisticStreamUnsafe();
+
+                            // Obtain the stream position as when transaction context last
+                            // remembered it.
+                            long streamReadPosition = getKnownStreamPosition()
+                                    .getOrDefault(proxy.getStreamID(), getSnapshotTimestamp());
+
                             return (
-                                getWriteSetEntrySize(proxy.getStreamID()) == 0 && // No updates
-                                        // And at the correct timestamp
-                                        o.getVersionUnsafe() == getSnapshotTimestamp()
-                                        && (stream == null
-                                            || stream.isStreamCurrentContextThreadCurrentContext())
-                            ); },
+                                    getWriteSetEntrySize(proxy.getStreamID()) == 0 /* No updates. */
+                                    && o.getVersionUnsafe() == streamReadPosition /* Match timestamp. */
+                                    && (stream == null || stream.isStreamCurrentContextThreadCurrentContext())
```

Comment:
Actually, if ``isStreamCurrentContextThreadCurrentContext()`` is true, and ``getWriteSetEntrySize(proxy.getStreamID()) == stream.pos()``, then we don't need to sync either

Reply:
I think it should be: getWriteSetEntrySize(proxy.getStreamID()) == **stream.pos()+1**?

Reply:
Yes, it should